### DISCUSSION TITLE
react-native-dotenv in react native expo working for ios and android but NOT web

### DISCUSSION THREAD

**Discussion by totallytotallyamazing** [goatandsheep/react-native-dotenv#407] [General]
Using react-native, "expo": "^46.0.14" in macOS Monterey 12.6.3, 
I installed "react-native-dotenv" in my react native expo app and it's working fine for ios and android but it's NOT working in my Web Development Build. 
Web does not see path: 
`import { EXPO_CLIENT_ID, ANDROID_CLIENT_ID, IOS_CLIENT_ID, WEB_CLIENT_ID } from "@env"; ` 
and reads, `console.log(process.env.EXPO_CLIENT_ID);` as `undefined`.
I installed `yarn add react-native-dotenv`
Created and put the following 2 files, in the root of my directory, see below: 
 `// .babelrc`
```JSON
{
    "plugins": [
        ["module:react-native-dotenv", {
            "envName": "APP_ENV",
            "moduleName": "@env",
            "path": ".env",
            "blocklist": null,
            "allowlist": null,
            "safe": false,
            "allowUndefined": true,
            "verbose": false
        }]
    ]
}
```

 `// .env`
```
EXPO_CLIENT_ID=SomeRandomNumber
ANDROID_CLIENT_ID=SomeRandomNumber
IOS_CLIENT_ID=SomeRandomNumber
WEB_CLIENT_ID=SomeRandomNumber
```
on https://www.npmjs.com/package/react-native-dotenv the claim is made:
Introduction
This babel plugin lets you inject your environment variables into your Javascript environment using dotenv for multiple environments. It is best suited for use with **react native and works with all flavors including web.**
Any thoughts, comments, or suggestions would be greatly appreciated!
Derek

**Comment by goatandsheep** [comment]
Although this is heavily tested on Android and iOS, I do most of my tests with web. Sorry it's not working for you! Let's figure out why that's the case.

* You don't need to set the values to the defaults, but it can't hurt!
* reinstall as devDependency `yarn add react-native-dotenv -D`
* try `console.log(EXPO_CLIENT_ID);`

**Reply by totallytotallyamazing** [reply]
Thank you for getting back to me! 
When you say, "You don't need to set the values to the defaults, but it can't hurt!", are you referring to my .babelrc file:
```JSON
{
    "plugins": [
        ["module:react-native-dotenv", {
            "envName": "APP_ENV",
            "moduleName": "@env",
            "path": ".env",
            "blocklist": null,
            "allowlist": null,
            "safe": false,
            "allowUndefined": true,
            "verbose": false
        }]
    ]
}
```
If so, can you suggest exactly which defaults I should try?

**Reply by goatandsheep** [reply]
the defaults don't need setting :)

```javascript
{
    "plugins": [
        ["module:react-native-dotenv"]
    ]
}
```

**Reply by totallytotallyamazing** [reply]
I tried all of your suggestions but still not working in Web.
•**Does Webpack require bootstrap by any chance?**


Here's my console errors:
```
Uncaught Error: Cannot find module '@env'
    at webpackMissingModule (firebase.js:34:1)
    at ./hooks/useAuth.js (firebase.js:34:1)
    at __webpack_require__ (bootstrap:789:1)
    at fn (bootstrap:100:1)
    at ./screens/HomeScreen.js (ChatScreen.js:28:1)
    at __webpack_require__ (bootstrap:789:1)
    at fn (bootstrap:100:1)
    at ./StackNavigator.js (StackNavigator.js:1:1)
    at __webpack_require__ (bootstrap:789:1)
    at fn (bootstrap:100:1)
    at ./App.js (App.js:1:1)
    at __webpack_require__ (bootstrap:789:1)
    at fn (bootstrap:100:1)
    at ./node_modules/expo/AppEntry.js (AppEntry.js:1:1)
    at __webpack_require__ (bootstrap:789:1)
    at fn (bootstrap:100:1)
    at 0 (PhotoScreen.js:25:1)
    at __webpack_require__ (bootstrap:789:1)
    at bootstrap:856:1
    at bootstrap:856:1
webpackMissingModule @ firebase.js:34
./hooks/useAuth.js @ firebase.js:34
__webpack_require__ @ bootstrap:789
fn @ bootstrap:100
./screens/HomeScreen.js @ ChatScreen.js:28
__webpack_require__ @ bootstrap:789
fn @ bootstrap:100
./StackNavigator.js @ StackNavigator.js:1
__webpack_require__ @ bootstrap:789
fn @ bootstrap:100
./App.js @ App.js:1
__webpack_require__ @ bootstrap:789
fn @ bootstrap:100
./node_modules/expo/AppEntry.js @ AppEntry.js:1
__webpack_require__ @ bootstrap:789
fn @ bootstrap:100
0 @ PhotoScreen.js:25
__webpack_require__ @ bootstrap:789
(anonymous) @ bootstrap:856
(anonymous) @ bootstrap:856
VM173:2 Uncaught ReferenceError: process is not defined
    at 4043 (<anonymous>:2:13168)
    at r (<anonymous>:2:306599)
    at 8048 (<anonymous>:2:9496)
    at r (<anonymous>:2:306599)
    at 8641 (<anonymous>:2:1379)
    at r (<anonymous>:2:306599)
    at <anonymous>:2:315627
    at <anonymous>:2:324225
    at <anonymous>:2:324229
    at e.onload (index.js:1:1)
4043 @ VM173:2
r @ VM173:2
8048 @ VM173:2
r @ VM173:2
8641 @ VM173:2
r @ VM173:2
(anonymous) @ VM173:2
(anonymous) @ VM173:2
(anonymous) @ VM173:2
e.onload @ index.js:1
be @ index.js:1
de @ index.js:1
handleErrors @ webpackHotDevClient.js:169
./node_modules/react-dev-utils/webpackHotDevClient.js.connection.onmessage @ webpackHotDevClient.js:213
load (async)
be @ index.js:1
de @ index.js:1
handleErrors @ webpackHotDevClient.js:169
./node_modules/react-dev-utils/webpackHotDevClient.js.connection.onmessage @ webpackHotDevClient.js:213
index.js:1 ./hooks/useAuth.js:10
Module not found: Can't resolve '@env'
   8 | import { GoogleAuthProvider, onAuthStateChanged, signInWithCredential, signOut } from 'firebase/auth';
   9 | // import Constants from 'expo-constants';
> 10 | import { EXPO_CLIENT_ID, ANDROID_CLIENT_ID, IOS_CLIENT_ID, WEB_CLIENT_ID, } from '@env';
  11 | // import { EXPO_CLIENT_ID, ANDROID_CLIENT_ID, IOS_CLIENT_ID, WEB_CLIENT_ID, } from '../web_client_id.js';
  12 |
  13 | // Below No longer needed after adding "process.env." to .env vars
```

**Reply by goatandsheep** [reply]
there is no chance bootstrap is a dependency :)

I would recommend that while expo withholds upgrading their babel (https://github.com/expo/expo-cli/pull/4593), you use `babel.config.js` and set `api.cache(false)`

**Reply by totallytotallyamazing** [reply]
Ok - no bootstrap... 
I'm reluctant to upgrade babel and think my versions are high enough (not sure, but basing that off the link you just sent).
```JSON
"dependencies": {
        "@babel/core": "7.9.0",
        "babel-loader": "8.1.0",
```
Are my versions high enough?

I've tried `api.cache(false)` and `api.cache(true)` in my `babel.config.js` to no avail. Also as a side note, the only way I have any success with getting this working is to utilize the `.babelrc` file and have tried different configurations to `babel.config.js` for pref changes but that never worked for me.

**Reply by totallytotallyamazing** [reply]
This config is working for ios and web and I think android Expo Go as well but it's currently giving me blue screen "Something went wrong." errors, which sometimes happen and are usually resolved by just waiting them out...

I removed this:
```JS
// import { EXPO_CLIENT_ID, ANDROID_CLIENT_ID, IOS_CLIENT_ID, WEB_CLIENT_ID } from '@env';
```
And changed by adding `process.env`:
```JS
console.log(process.env.EXPO_CLIENT_ID);

    const [request, response, promptAsync] = Google.useAuthRequest({
        expoClientId: `${process.env.EXPO_CLIENT_ID}.apps.googleusercontent.com`,
        androidClientId: `${process.env.ANDROID_CLIENT_ID}.apps.googleusercontent.com`,
        iosClientId: `${process.env.IOS_CLIENT_ID}.apps.googleusercontent.com`,
        webClientId: `${process.env.WEB_CLIENT_ID}.apps.googleusercontent.com`,
        redirectUri: makeRedirectUri({
            scheme: 'host.exp.exponent',
            useProxy: true,
            native: Platform.select({
                android: `com.googleusercontent.apps.${process.env.ANDROID_CLIENT_ID}:/oauth2redirect`,
                default: undefined
            })
        }),
        scopes: ['profile', 'email'],
        permissions: ['public_profile', 'email', 'gender', 'location']
    });
```
Do you have any thoughts on this setup? Will it interfere with things when I go to hide my APIs with say, express on the backend or some other backend solution?

**Reply by totallytotallyamazing** [reply]
Hey Kemal,
Do you have any response to my questions from 3 weeks ago?
Thank you for the help you did provide already but I'm still left a bit unsure about my use of react-native-dotenv.
Thanks,
Derek

**Reply by goatandsheep** [reply]
Hey Derek, sorry for the late response. did you install the lib as a devDependency or regular dependency?

**Reply by totallytotallyamazing** [reply]
Hey Kemal, I recently upgraded to expo sdk 49 from 47 and don't require a dotenv plugin any longer, not sure if I needed it with 47 either, since my problem is fixed there was no need to do backwards research just for knowledge.

Thank you for the help you could offer me though.
Derek
